pennylane/optimize/qnspsa.py

```diff
@@ -0,0 +1,407 @@
+            warnings.warn(
+                "step_and_cost() instead of step() is called when "
+                "blocking is turned on, as the step-wise loss value "
+                "is required by the algorithm.",
+                stacklevel=2,
+            )
+            return self.step_and_cost(cost, *args, **kwargs)[0]
+
+        return self._step_core(cost, args, kwargs)
+
+    def step_and_cost(self, cost, *args, **kwargs):
+        r"""Update trainable parameters with one step of the optimizer and return
+        the corresponding objective function value after the step.
+
+        Args:
+            cost (QNode): the QNode wrapper for the objective function for optimization
+            *args : variable length argument list for qnode
+            **kwargs : variable length of keyword arguments for the qnode
+
+        Returns:
+            (np.array, float): the new variable values :math:`x^{(t+1)}` and the objective
+            function output prior to the step
+        """
+        params_next = self._step_core(cost, args, kwargs)
+
+        if not self.blocking:
+            loss_curr = cost(*args, **kwargs)
+            return params_next, loss_curr
+
+        params_next, loss_curr = self._apply_blocking(cost, args, kwargs, params_next)
+        return params_next, loss_curr
+
+    def _step_core(self, cost, args, kwargs):
+        all_grad_tapes = []
+        all_grad_dirs = []
+        all_metric_tapes = []
+        all_tensor_dirs = []
+        for i in range(self.resamplings):
+            # grad_tapes contains 2 tapes for the gradient estimation
+            grad_tapes, grad_dirs = self._get_spsa_grad_tapes(cost, args, kwargs)
+            # metric_tapes contains 4 tapes for tensor estimation
+            metric_tapes, tensor_dirs = self._get_tensor_tapes(cost, args, kwargs)
+            all_grad_tapes += grad_tapes
+            all_metric_tapes += metric_tapes
+            all_grad_dirs.append(grad_dirs)
+            all_tensor_dirs.append(tensor_dirs)
+
+        raw_results = qml.execute(all_grad_tapes + all_metric_tapes, cost.device, None)
+        grads = [
+            self._post_process_grad(raw_results[2 * i : 2 * i + 2], all_grad_dirs[i])
+            for i in range(self.resamplings)
+        ]
+        grads = np.array(grads)
+        metric_tensors = [
+            self._post_process_tensor(
+                raw_results[2 * self.resamplings + 4 * i : 2 * self.resamplings + 4 * i + 4],
+                all_tensor_dirs[i],
+            )
+            for i in range(self.resamplings)
+        ]
+        metric_tensors = np.array(metric_tensors)
+        grad_avg = np.mean(grads, axis=0)
+        tensor_avg = np.mean(metric_tensors, axis=0)
+
+        self._update_tensor(tensor_avg)
+        params_next = self._get_next_params(args, grad_avg)
+
+        if len(params_next) == 1:
+            return params_next[0]
+        return params_next
+
+    def _post_process_grad(self, grad_raw_results, grad_dirs):
```

Comment:
Would be nice to have a one-sentence docstring here as well, with short descriptions of what the arguments are without having to look for usages of this function.